scripts/ci/dag.py

```diff
@@ -2,11 +2,38 @@
 
 import time
 from concurrent.futures import ThreadPoolExecutor
+from math import floor
 from multiprocessing import Event, cpu_count
 from multiprocessing.synchronize import Event as EventClass
 from queue import Empty, Queue
 from typing import Callable, Generic, Hashable, TypeVar
 
+
+class RateLimiter:
+    """
+    Burst rate limiter.
+
+    Starts at `max_tokens`, and refills one token every `refill_interval_sec / max_tokens`.
+    """
+
+    def __init__(self, max_tokens: int, refill_interval_sec: float):
+        self.start_tokens = max_tokens
+        self.tokens_per_second = max_tokens / refill_interval_sec
+        self.start_time = time.time()
+        self.used_tokens = 0
+
+    def get(self) -> bool:
```

Comment:
I think this does the right thing now? The way I understand it, it keeps track of the number of used tokens and compares that against the total number of tokens across the whole lifetime of the rate limiter. I think the snippet you posted wasn't taking into account the rate at which tokens should be refreshed, and always did one per second, which would be too many for crates.io (it would exceed 30/s at 60s). Still not sure if this is correct, because math-heavy code is a challenge for me!

Reply:
Here's the crates.io rate limiter: https://github.com/rust-lang/crates.io/blob/main/src/rate_limiter.rs

Reply:
Okay, one more try, I think this is still wrong, and you were right yet again. They _do_ refill one per second.

Reply:
If you are aiming to mimic https://github.com/rust-lang/crates.io/blob/main/src/rate_limiter.rs then at least add a link to in the code for reference.

But it doesn't really have to be perfect does it? crates.io tells us how much to sleep if we are eager so we will self-adjust.

Anyways: I read their code, and what they call `burst` seems to be the initial number of tokens and the maximum ever available. You call this `max_tokens`. This seems to be 30 for our use-case (mostly updating existing crates).

Then it looks like they add ONE token per 60 seconds. NOT `burst` amount, at least not what I can see. So I think your code should read `self.tokens_per_second = 1.0 / refill_interval_sec`. Then it should be good to go!

Reply:
Thanks! Updated.